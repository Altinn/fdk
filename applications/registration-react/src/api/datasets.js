import axios from 'axios';

const datasetUrlBase = '/catalogs';

export const getDatasetByURI = async uri => {
  const url = `/search-api/datasets/byuri?uri=${uri}`;

  const response = await axios
    .get(url)
    .catch(e => console.log(JSON.stringify(e))); // eslint-disable-line no-console

  return response && response.data;
};

export const getDatasetByTitlePrefix = async (
  title = '',
  orgPath,
  returnFields
) => {
  const queryParams = `title=${title}${orgPath ? `&orgPath=${orgPath}` : ''}${
    returnFields ? `&returnfields=${returnFields}` : ''
  }`;

  const url = `/search-api/datasets?${queryParams}`;

  const response = await axios
    .get(url)
    .catch(e => console.log(JSON.stringify(e))); // eslint-disable-line no-console

  return response && response.data;
};

export const getDatasetById = async (id, catalogId) => {
  const url = `/catalogs/${catalogId}/datasets/${id}`;

  const response = await axios
    .get(url)
    .catch(e => console.log(JSON.stringify(e))); // eslint-disable-line no-console

  return response && response.data;
};

export const patchDataset = (catalogId, datasetId, header, body) =>
  axios
    .patch(`${datasetUrlBase}/${catalogId}/datasets/${datasetId}`, body, header)
    .then(response => response);
