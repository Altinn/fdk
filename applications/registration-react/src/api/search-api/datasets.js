import { getConfig } from '../../config';
import { searchApiGet } from './host';

export const getDatasetByURI = uri =>
  searchApiGet({
    url: `${getConfig().datasetApi.host}/api/datasets/byuri`,
    params: { uri },
    authorization: getConfig().datasetApi.authorization
  });

export const searchDatasets = ({ title, orgPath, returnFields }) =>
  searchApiGet({
    url: `${getConfig().datasetApi.host}/api/datasets`,
    params: { title, orgPath, returnFields },
    authorization: getConfig().datasetApi.authorization
  });
