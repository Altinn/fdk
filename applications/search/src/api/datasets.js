import axios from 'axios';
import qs from 'qs';

const datasetsUrlBase = '/datasets';

const searchAggregations =
  'accessRightsCount,theme_count,orgPath,catalogs,provenanceCount,firstHarvested,missingFirstHarvested,lastChanged,missingLastChanged,spatial,opendata';

export const datasetsSearchUrl = query =>
  `${datasetsUrlBase}${qs.stringify(
    { ...query, aggregations: searchAggregations },
    { addQueryPrefix: true }
  )}`;

export const getDataset = id =>
  axios
    .get(`${datasetsUrlBase}/${id}`)
    .then(response => response.data)
    .catch(e => console.log(JSON.stringify(e))); // eslint-disable-line no-console

export const getDatasetByURI = uri =>
  axios
    .get(`${datasetsUrlBase}/byuri?uri=${uri}`)
    .then(response => response.data)
    .catch(e => console.log(JSON.stringify(e))); // eslint-disable-line no-console
