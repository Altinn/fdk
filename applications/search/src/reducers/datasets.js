import _ from 'lodash';
import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../constants/ActionTypes';

export default function datasets(
  state = {
    isFetchingDatasets: false,
    datasetItems: null,
    publisherCountItems: null
  },
  action
) {
  switch (action.type) {
    case DATASETS_REQUEST: {
      return {
        ...state,
        isFetchingDatasets: true
      };
    }
    case DATASETS_SUCCESS: {
      const orgs = action.response.data.aggregations.orgPath.buckets;
      const flat = _(orgs).forEach(f => {
        const filteredOrgs = _(orgs)
          .filter(g => g.key.substring(0, g.key.lastIndexOf('/')) === f.key)
          .value();
        filteredOrgs.forEach(item => {
          const retVal = item;
          retVal.hasParent = true;
          return retVal;
        });
        const retVal = f;
        retVal.children = filteredOrgs;
        return retVal;
      });
      const resultArray = _(flat)
        .filter(f => !f.hasParent)
        .value();

      const objFromArray = action.response.data.aggregations.subjectsCount.buckets.reduce(
        (accumulator, current) => {
          accumulator[current.key] = current; // eslint-disable-line no-param-reassign
          return accumulator;
        },
        {}
      );
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: action.response.data,
        publisherCountItems: resultArray,
        subjectsCountItems: objFromArray
      };
    }
    case DATASETS_FAILURE: {
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: null,
        publisherCountItems: null
      };
    }
    default:
      return state;
  }
}

export const getDatasetById = (datasets, id) => {
  if (datasets && datasets.datasetItems && datasets.datasetItems.hits) {
    return datasets.datasetItems.hits.hits.filter(
      dataset => dataset._source.id === id
    );
  }
  return null;
};
