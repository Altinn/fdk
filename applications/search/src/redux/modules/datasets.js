import _ from 'lodash';
import { normalizeAggregations } from '../../lib/normalizeAggregations';
import { fetchActions } from '../fetchActions';

export const DATASETS_REQUEST = 'DATASETS_REQUEST';
export const DATASETS_SUCCESS = 'DATASETS_SUCCESS';
export const DATASETS_FAILURE = 'DATASETS_FAILURE';

function shouldFetch(metaState, query) {
  const threshold = 60 * 100; // seconds
  return (
    !metaState ||
    (!metaState.isFetching &&
      ((metaState.lastFetch || 0) < Date.now() - threshold ||
        metaState.cachedQuery !== query))
  );
}

export function fetchDatasetsIfNeededAction(query) {
  return (dispatch, getState) =>
    shouldFetch(_.get(getState(), ['datasets', 'meta']), query) &&
    dispatch(
      fetchActions(`/datasets?${query}`, [
        DATASETS_REQUEST,
        { type: DATASETS_SUCCESS, meta: { query } },
        DATASETS_FAILURE
      ])
    );
}

const initialState = {};

export function datasets(state = initialState, action) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        meta: {
          isFetching: true,
          lastFetch: null,
          cachedQuery: null
        }
      };
    case DATASETS_SUCCESS:
      return {
        ...state,
        datasetItems: _.get(action.payload, ['hits', 'hits']),
        datasetAggregations: _.get(normalizeAggregations(action.payload), [
          'aggregations'
        ]),
        datasetTotal: _.get(action.payload, ['hits', 'total']),
        meta: {
          isFetching: false,
          lastFetch: Date.now(),
          cachedQuery: action.meta.query
        }
      };
    case DATASETS_FAILURE:
      return {
        ...state,
        datasetItems: null,
        datasetAggregations: null,
        datasetTotal: null,
        meta: {
          isFetching: false,
          lastFetch: null,
          cachedQuery: null
        }
      };
    default:
      return state;
  }
}
