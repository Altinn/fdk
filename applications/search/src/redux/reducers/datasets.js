import _ from 'lodash';
import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../ActionTypes';

export default function datasets(
  state = {
    isFetching: false,
    datasetItems: null,
    publisherCountItems: null
  },
  action
) {
  switch (action.type) {
    case DATASETS_REQUEST: {
      return {
        ...state,
        isFetching: true
      };
    }
    case DATASETS_SUCCESS: {
      const orgs = action.payload.aggregations.orgPath.buckets;
      const flat = _(orgs).forEach(f => {
        const filteredOrgs = _(orgs)
          .filter(g => g.key.substring(0, g.key.lastIndexOf('/')) === f.key)
          .value();
        filteredOrgs.forEach(item => {
          const retVal = item;
          retVal.hasParent = true;
          return retVal;
        });
        const retVal = f;
        retVal.children = filteredOrgs;
        return retVal;
      });
      const resultArray = _(flat)
        .filter(f => !f.hasParent)
        .value();

      const objFromArray = action.payload.aggregations.subjectsCount.buckets.reduce(
        (accumulator, current) => {
          accumulator[current.key] = current; // eslint-disable-line no-param-reassign
          return accumulator;
        },
        {}
      );
      return {
        ...state,
        isFetching: false,
        datasetItems: action.payload,
        publisherCountItems: resultArray,
        subjectsCountItems: objFromArray
      };
    }
    case DATASETS_FAILURE: {
      return {
        ...state,
        isFetching: false,
        datasetItems: null,
        publisherCountItems: null
      };
    }
    default:
      return state;
  }
}

export const getDatasetById = (datasets, id) => {
  if (datasets && datasets.datasetItems && datasets.datasetItems.hits) {
    return datasets.datasetItems.hits.hits.filter(
      dataset => dataset._source.id === id
    );
  }
  return null;
};
