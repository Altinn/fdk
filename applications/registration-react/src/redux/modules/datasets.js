import _ from 'lodash';
import { fetchActions } from '../fetchActions';
import { registrationApi } from '../../api/registration-api';

export const DATASETS_REQUEST = 'DATASETS_REQUEST';
export const DATASETS_SUCCESS = 'DATASETS_SUCCESS';
export const DATASETS_FAILURE = 'DATASETS_FAILURE';
export const DATASETS_ITEM_DELETE = 'DATASETS_ITEM_DELETE';
export const DATASET_SUCCESS = 'DATASET_SUCCESS';

function shouldFetch(metaState) {
  const threshold = 60 * 1000; // seconds
  return (
    !metaState ||
    (!metaState.isFetching &&
      (metaState.lastFetch || 0) < Date.now() - threshold)
  );
}

export function fetchDatasetsIfNeeded(catalogId) {
  return (dispatch, getState) =>
    shouldFetch(_.get(getState(), ['datasets', catalogId, 'meta'])) &&
    dispatch(
      fetchActions(`/catalogs/${catalogId}/datasets`, [
        { type: DATASETS_REQUEST, meta: { catalogId } },
        { type: DATASETS_SUCCESS, meta: { catalogId } },
        { type: DATASETS_FAILURE, meta: { catalogId } }
      ])
    );
}

export const deleteDatasetItemAction = (catalogId, datasetId) => ({
  type: DATASETS_ITEM_DELETE,
  catalogId,
  datasetId
});

export const datasetSuccessAction = payload => ({
  type: DATASET_SUCCESS,
  payload
});

export const deleteDatasetAction = (catalogId, datasetId) => async dispatch => {
  await registrationApi.deleteDataset(catalogId, datasetId);
  dispatch(deleteDatasetItemAction(catalogId, datasetId));
};

const initialState = {};

export default function datasets(state = initialState, action) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: true,
            lastFetch: null
          }
        }
      };
    case DATASETS_SUCCESS: {
      const objFromArray = _.get(action.payload, [
        '_embedded',
        'datasets'
      ]).reduce((accumulator, current) => {
        accumulator[current.id] = current; // eslint-disable-line no-param-reassign
        return accumulator;
      }, {});
      return {
        ...state,
        [action.meta.catalogId]: {
          items: objFromArray,
          meta: {
            isFetching: false,
            lastFetch: Date.now()
          }
        }
      };
    }
    case DATASETS_FAILURE:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: false,
            lastFetch: null
          }
        }
      };
    case DATASET_SUCCESS: {
      const items = _.get(state, [action.payload.catalogId, 'items']);
      return {
        ...state,
        [action.payload.catalogId]: {
          items: {
            ...items,
            [action.payload.id]: action.payload
          }
        }
      };
    }
    case DATASETS_ITEM_DELETE:
      return {
        ...state,
        [action.catalogId]: {
          ..._.omit(state, [action.catalogId, 'items'], action.datasetId)
        }
      };
    default:
      return state;
  }
}

export const getDatasetItemsByCatalogId = (datasets, catalogId) =>
  Object.values(_.get(datasets, [catalogId, 'items'], {}));

export const getDatasetItemsCount = (datasets, catalogId) =>
  getDatasetItemsByCatalogId(datasets, catalogId).length;

export const getDatasetItemByDatasetiId = (datasets, catalogId, id) =>
  _.get(datasets, [catalogId, 'items', id]);
