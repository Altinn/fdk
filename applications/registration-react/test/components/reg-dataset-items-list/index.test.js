import React from 'react';
import { shallow } from 'enzyme';
import DatasetItemsList from '../../../src/components/reg-dataset-items-list';
import datasets from '../../fixtures/datasets';

let defaultProps, wrapper, refreshSession, toggle;

beforeEach(() => {
  defaultProps = {
    catalogId: '123'
  };
  wrapper = shallow(<DatasetItemsList {...defaultProps} />);
});

test('should render DatasetItemsList correctly with missing datasetItems', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render DatasetItemsList correctly with datasetItems', () => {
  const { datasetItems } = datasets;
  wrapper.setProps({
    datasetItems
  })
  expect(wrapper).toMatchSnapshot();
});
