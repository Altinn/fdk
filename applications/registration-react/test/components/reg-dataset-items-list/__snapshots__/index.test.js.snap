// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DatasetItemsList correctly with datasetItems 1`] = `
<div>
  <div
    className="d-flex mb-3"
  >
    <button
      className="fdk-button fdk-button-cta"
      onClick={[Function]}
    >
      <i
        className="fa fa-plus fdk-color0 mr-2"
      />
      Legg til ny datasettbeskrivelse
    </button>
    <button
      className="ml-2 transparentButton"
      onClick={[Function]}
    >
      <i
        className="fa fa-plus fdk-color1 mr-2"
      />
      Importer datasett
    </button>
  </div>
  <div
    className="fdk-datasets-list-header d-flex"
  >
    <div
      className="d-flex align-items-center w-75"
    >
      <span
        className="header-item mr-1"
      >
        Tittel p책 datasett
      </span>
      <SortButtons
        field="title"
        onSortField={[Function]}
        sortField=""
        sortType=""
      />
    </div>
    <div
      className="d-flex align-items-center w-25"
    >
      <span
        className="header-item mr-1"
      >
        Status
      </span>
      <SortButtons
        field="registrationStatus"
        onSortField={[Function]}
        sortField=""
        sortType=""
      />
    </div>
  </div>
  <DatasetItemsListItem
    catalogId="123"
    item={
      Object {
        "_lastModified": "2018-05-07T10:36:20.388+0000",
        "accessRights": Object {
          "prefLabel": Object {},
          "uri": "http://publications.europa.eu/resource/authority/access-right/PUBLIC",
        },
        "catalogId": "910244132",
        "description": Object {
          "nb": "tester",
        },
        "id": "084f747f-04a9-460e-b07e-ee74c60a2086",
        "landingPage": Array [
          "",
        ],
        "legalBasisForAccess": Array [
          Object {
            "extraType": null,
            "prefLabel": Object {
              "nb": "",
            },
            "uri": "",
          },
        ],
        "legalBasisForProcessing": Array [
          Object {
            "extraType": null,
            "prefLabel": Object {
              "nb": "",
            },
            "uri": "",
          },
        ],
        "legalBasisForRestriction": Array [
          Object {
            "extraType": null,
            "prefLabel": Object {
              "nb": "",
            },
            "uri": "",
          },
        ],
        "objective": Object {
          "nb": "",
        },
        "publisher": Object {
          "name": "RAMSUND OG ROGNAN REVISJON",
          "uri": "http://data.brreg.no/enhetsregisteret/enhet/910244132",
        },
        "registrationStatus": "PUBLISH",
        "theme": Array [
          Object {
            "title": Object {
              "nb": "",
            },
            "uri": "",
          },
          Object {
            "title": Object {
              "nb": "Jordbruk, fiskeri, skogbruk og mat",
            },
            "uri": "http://publications.europa.eu/resource/authority/data-theme/AGRI",
          },
        ],
        "title": Object {
          "nb": "test ois",
        },
        "uri": "http://brreg.no/catalogs/910244132/datasets/084f747f-04a9-460e-b07e-ee74c60a2086",
      }
    }
    key="084f747f-04a9-460e-b07e-ee74c60a2086"
  />
  <AppImportModal
    body="DCAT-modellen som blir importert m책 inneholde en katalog som har samme organisasjonsnr som katalogen du redigerer. Kun datasett som er med i denne katalogen vil bli importert."
    btnCancel="Avbryt"
    btnConfirm="Importer"
    className={null}
    handleAction={[Function]}
    modal={false}
    setURL="Angi url til DCAT ressurs:"
    title="Importer datasett fra DCAT RDF ressurs."
    toggle={[Function]}
  />
</div>
`;

exports[`should render DatasetItemsList correctly with missing datasetItems 1`] = `
<div>
  <div
    className="d-flex mb-3"
  >
    <button
      className="fdk-button fdk-button-cta"
      onClick={[Function]}
    >
      <i
        className="fa fa-plus fdk-color0 mr-2"
      />
      Legg til ny datasettbeskrivelse
    </button>
    <button
      className="ml-2 transparentButton"
      onClick={[Function]}
    >
      <i
        className="fa fa-plus fdk-color1 mr-2"
      />
      Importer datasett
    </button>
  </div>
  <div
    className="fdk-datasets-list-header d-flex"
  >
    <div
      className="d-flex align-items-center w-75"
    >
      <span
        className="header-item mr-1"
      >
        Tittel p책 datasett
      </span>
      <SortButtons
        field="title"
        onSortField={[Function]}
        sortField=""
        sortType=""
      />
    </div>
    <div
      className="d-flex align-items-center w-25"
    >
      <span
        className="header-item mr-1"
      >
        Status
      </span>
      <SortButtons
        field="registrationStatus"
        onSortField={[Function]}
        sortField=""
        sortType=""
      />
    </div>
  </div>
  <div
    className="fdk-datasets-list-item d-flex"
  >
    <span
      className="fdk-text-size-small fdk-color2"
    >
      Ingen datasettbeskrivelse registrert
    </span>
  </div>
  <AppImportModal
    body="DCAT-modellen som blir importert m책 inneholde en katalog som har samme organisasjonsnr som katalogen du redigerer. Kun datasett som er med i denne katalogen vil bli importert."
    btnCancel="Avbryt"
    btnConfirm="Importer"
    className={null}
    handleAction={[Function]}
    modal={false}
    setURL="Angi url til DCAT ressurs:"
    title="Importer datasett fra DCAT RDF ressurs."
    toggle={[Function]}
  />
</div>
`;
